clients/drcachesim/drpt2trace/pt2ir.h

```diff
@@ -0,0 +1,263 @@
+ * DAMAGE.
+ */
+
+/* pt2ir: convert a PT raw trace to DynamoRIO's IR format. */
+
+#ifndef _PT2IR_H_
+#define _PT2IR_H_ 1
+
+/**
+ * @file drpt2trace/pt2ir.h
+ * @brief Offline PT raw trace converter.
+ */
+
+#include <string>
+#include <vector>
+#define DR_FAST_IR 1
+#include "dr_api.h"
+
+#ifndef IN
+#    define IN // nothing
+#endif
+#ifndef OUT
+#    define OUT // nothing
+#endif
+#ifndef INOUT
+#    define INOUT // nothing
+#endif
+
+/**
+ * The type of pt2ir_t::convert() return value.
+ */
+enum pt2ir_convert_status_t {
+    PT2IR_CONV_SUCCESS = 0,       /**< The conversion process is successful. */
+    PT2IR_CONV_SYNC_PACKET_ERROR, /**< The conversion process ends with a failure to sync
+                                   *   to the PSB packet.
+                                   */
+    PT2IR_CONV_HANDLE_SIDEBAND_EVENT_ERROR, /**< The conversion process ends with a
+                                             *   failure to handle a perf event.
+                                             */
+    PT2IR_CONV_GET_PENDING_EVENT_ERROR, /**< The conversion process ends with a failure to
+                                         *   get the pending event.
+                                         */
+    PT2IR_CONV_SET_IMAGE_ERROR, /**< The conversion process ends with a failure to set the
+                                 *   new image.
+                                 */
+    PT2IR_CONV_DECODE_NEXT_INSTR_ERROR /**< The conversion process ends with a failure to
+                                        *   decode the next intruction.
+                                        */
+};
+
+/**
+ * The struct pt2ir_config_t is a collection of one PT trace's configurations. drpt2trace
+ * and raw2trace can use it to construct pt2ir_t.
+ * XXX: Multiple PT raw traces for a process will share one kcore dump file. We may need a
+ * kernel image manager to avoid loading the same dump file many times.
+ */
+struct pt2ir_config_t {
+    /* The class pt2ir_t does not want to expose libipt to the upper layer. So we redefine
+     * some configurations of libipt decoder in pt_config. The libipt pt decoder requires
+     * these parameters. We can get these parameters by running
+     * libipt/scirpts/perf-get-opts.bash.
+     */
+    struct {
+        /* A cpu identifier.*/
+        struct {
+            /* The vendor of the cpu(0 for unknown, 1 for Intel). */
+            uint8_t vendor;
+
+            /* The cpu family, model and stepping */
+            uint16_t family;
+            uint8_t model;
+            uint8_t stepping;
+        } cpu;
+
+        /* cpuid_0x15_eax and cpuid_0x15_ebx represent the CTC frequency*/
+        uint32_t cpuid_0x15_eax;
+        uint32_t cpuid_0x15_ebx;
+
+        /* The MTC frequency. */
+        uint8_t mtc_freq;
+
+        /* The nominal frequency. */
+        uint8_t nom_freq;
+    } pt_config; /**< The libipt config of PT raw trace. We can get these parameters by
```

Comment:
Seems clearer to put this Doxygen comment up before the struct.

Reply:
Done.